aptos/contract/aux/sources/stake.move

```diff
@@ -0,0 +1,1399 @@
+
+        pool.authority = new_authority;
+
+        event::emit_event<ModifyPoolEvent<S, R>>(
+            &mut pools.modify_pool_events,
+            ModifyPoolEvent {
+                pool_id: id,
+                authority: pool.authority,
+                start_time: pool.start_time,
+                end_time: pool.end_time,
+                reward_remaining: pool.reward_remaining,
+                total_amount_staked: coin::value(&pool.stake),
+                acc_reward_per_share: pool.acc_reward_per_share
+            }
+        );
+    }
+
+    /// Modify incentive pool parameters.
+    /// Only the pool's authority can modify parameters.
+    public entry fun modify_pool<S, R>(
+        sender: &signer,
+        id: u64,
+        reward_amount: u64,
+        reward_increase: bool,
+        time_amount_us: u64,
+        time_increase: bool,
+    ) acquires Pools {
+        assert!(exists<Pools<S, R>>(@aux), E_INCENTIVE_POOL_NOT_FOUND);
+        let pools = borrow_global_mut<Pools<S, R>>(@aux);
+        assert!(table::contains(&mut pools.pools, id), E_INCENTIVE_POOL_NOT_FOUND);
+        let pool = table::borrow_mut(&mut pools.pools, id);
+
+        let sender_addr = signer::address_of(sender);
+        assert!(sender_addr == pool.authority, E_NOT_AUTHORIZED);
+
+        let now = timestamp::now_microseconds();
+        update_pool(pool, now);
+
+        if (reward_amount > 0 && reward_increase) {
+            let reward = coin::withdraw<R>(sender, reward_amount);
+            coin::merge(&mut pool.reward, reward);
+            pool.reward_remaining = pool.reward_remaining + reward_amount;
+        } else if (reward_amount > 0 && !reward_increase) {
+            assert!(pool.reward_remaining >= reward_amount, E_INVALID_REWARD);
+            let reward = coin::extract(&mut pool.reward, reward_amount);
+            if (!coin::is_account_registered<R>(sender_addr)) {
+                coin::register<R>(sender);
+            };
+            coin::deposit(sender_addr, reward);
+            pool.reward_remaining = pool.reward_remaining - reward_amount;
+            if (pool.reward_remaining == 0) {
+                pool.end_time = now;
+            }
+        };
+        // the real balance of the reward coin should always be >= the virtual remaining balance
+        assert!(coin::value(&pool.reward) >= pool.reward_remaining, E_INTERNAL_ERROR);
+
+        if (time_amount_us > 0 && time_increase) {
+            // If the pool has expired, start the reward period from now
+            if (now >= pool.end_time) {
+                pool.start_time = now;
+                pool.end_time = now + time_amount_us;
+            } else {
+                pool.end_time = pool.end_time + time_amount_us;
+            }
+        // TODO: should we allow this?
+        } else if (time_amount_us > 0 && !time_increase) {
+            assert!(time_amount_us < pool.end_time, E_INVALID_DURATION);
+            pool.end_time = pool.end_time - time_amount_us;
```

Comment:
If the end time of the pool changes, do we need to update the acc_reward_per_share, etc.?

Reply:
That update will happen on the next `update_pool`. `acc_reward_per_share` represents the cumulative reward per share that has been paid so far. `update_pool()` is called at the beginning of `modify_pool()` to add the reward per share over the duration `(pool.last_update_time, now)` to `acc_reward_per_share`. The next time the pool is updated, it will add to `acc_reward_per_share` based on the update rate, so everyone staking over that duration receives that rate.